### PR TITLE
Upgrade to Gun2.0

### PR DESCRIPTION
**Pull request by belltoy**
Based on #55.

And fixes some var warnings in test suites.

### CODE REVIEW

**Comment by zhongwencool** [comment]
I don't have time to review this right now :(. 
 please help to review this @enjolras1205

**Comment by belltoy** [comment]
Gun 2.0 is finally out so the original PR #55 can upgrade to it directly. Or I can do this based on #55 if @enjolras1205 don't mind.

@zhongwencool Would you merge this and bump a new version on hex.pm? After reviews, off course.

**Comment by belltoy** [comment]
The original PR introduced some broken changes. So I will add some changelogs if those changes are acceptable.

**Comment by belltoy** [comment]
The changes from #55 seem good to me.

As per the discussions in #28, I upgrade eetcd version to 0.4.0 with a migration note attached to the README.

Fixes the outdated docs for the `eetcd:open/5,6`.

File: src/eetcd.erl [hunk 1]
```diff
@@ -11,19 +11,21 @@
...
     [string()],
     tcp | tls | ssl,
-    [gen_tcp:connect_option()] | [ssl:connect_option()]) ->
+    [gen_tcp:connect_option()],
+    [ssl:tls_client_option()]) ->
     {ok, pid()} | {error, any()}.
-open(Name, Hosts, Transport, TransportOpts) ->
-    open(Name, Hosts, [], Transport, TransportOpts).
+open(Name, Hosts, Transport, TcpOpts, TlsOpts) when is_atom(Transport) ->
+    open(Name, Hosts, [], Transport, TcpOpts, TlsOpts).
```
Comment (gilbertwong96) [inline src/eetcd.erl:24]: It is a little strange to put `TcpOps` and  `TlsOpts` together into one function.

- `TcpOpts` is associated with the transport `tcp`
- `TlsOpts` is associated with the transport `ssl` or `tls`

Maybe this way is a little better:

I think it is better to combine `Transport` and `TransportOpts` into one tuple and place it into the `Options`.

```erlang
-spec open(Name, Hosts, Options)  -> {ok, pid()} | {error, any()} when
    Name :: name(),
    Hosts :: [string()],
    Options :: [ {mode, connect_all | random} |
                        {name, string()} |
                        {password, string()}  |
                        {retry, non_neg_integer()} |
                        {retry_timeout, pos_integer()} |
                        {connect_timeout, timeout()} 
                        {transport, tcp | ssl | tls} |
                        {transport_opts, [gen_tcp:connect_option()] |  [ssl:tls_client_option()]}
                      ].
```

Comment (gilbertwong96) [reply src/eetcd.erl:24]: Regardless, this pull request will introduce a breaking change. I believe we can offer a simpler API in this case.

File: README.md [hunk 2]
```diff
@@ -289,3 +289,24 @@ Gen proto and client file
 ```erlang
 rebar3 etcd gen
 ```
+
+Migration from eetcd 0.3.6 to 0.4.x
+-----
+
+eetcd 0.4.x now dependents on Gun 2.0, which introduced some broken changes,
```
Comment (michaelklishin) [inline README.md:296]: You mean "breaking changes", not "broken" (incorrect) changes.

File: src/eetcd.erl (hunk 1 above)
Comment (belltoy) [reply src/eetcd.erl]: That makes sense. I have tried to make this change, and it looks good.

**Comment by zhongwencool** [comment]
Thanks a lot!

**Pull request merged**